Fixed #33143 -- Raised RuntimeWarning when performing import-time queries. (#16923)
django/django: docs/ref/applications.txt

        separate from the production settings, ``manage.py test`` would still
        execute some queries against your **production** database!

    .. versionchanged:: 5.0

        Django will raise a ``RuntimeWarning`` when you interact with the
        database before the app registry has been fully populated.

> 

Reviewer: Based on Carlton's comment in the ticket, ideally we'll explain or add a pointer about how the user can fix this warning.

Author: I've added a link to the troubleshooting section of the applications documentation to the warning message. I've also added a few paragraphs to the docs that reproduce the example from the [original ticket](https://code.djangoproject.com/ticket/33143), suggest using a `ModelChoiceField`, and point to Python's [warning filter](https://docs.python.org/3/library/warnings.html#the-warnings-filter) in the hopes that people could more easily discover the places in which they perform import-time queries.

Author: Removed the link to the troubleshooting section again because of @adamchainz's [comment](https://github.com/django/django/pull/16923#discussion_r1218613418) below.

    It must be called explicitly in other cases, for instance in plain Python
    scripts.

    .. versionchanged:: 5.0

        Raises a ``RuntimeWarning`` when apps interact with the database before

>         the app registry has been fully populated.

Reviewer 2: I don't think this note is needed, it's duplicative?

Author: Yes, but the two notes are somewhat far apart and are attached to two different methods. If I _had_ to choose one of the two instances, I would keep the one below `django.setup()` because `django.setup()` actually triggers the warning. The one below `AppConfig.ready()` just felt natural because it was right below the big fat warning block about execute SQL queries in `AppConfig.ready()`.

Author: I've removed the note blow `AppConfig.ready()` and kept the one below `django.setup()`

  :setting:`INSTALLED_APPS` to contain
  ``'django.contrib.admin.apps.SimpleAdminConfig'`` instead of
  ``'django.contrib.admin'``.

* ``RuntimeWarning: Accessing the database during app initialization is
  discouraged.`` This happens if one of your apps executes SQL queries in either
  the :meth:`AppConfig.ready` method or during import time. This is bad
  because it will happen during the startup of every management command.

  For example, a common mistake may be to populate the choices of a form field
  like this::

    class SomeForm(forms.Form):
        country = forms.ChoiceField(choices=[c.name for c in Country.objects.all()])

  In the example above, the ``Country.objects.all()`` query would be executed as
  soon as the module containing this form would be imported. To silence the
  warning, use a :class:`~django.forms.ModelChoiceField` instead::

    class SomeForm(forms.Form):
        country = forms.ModelChoiceField(queryset=Country.objects.all())

  To make it easier to find the code that triggered this warning, tell Python
  to :ref:`treat warnings as errors <python:warning-filter>` using, for example,

>   ``python -Werror manage.py shell`` which would give you a complete stack trace.

Reviewer 3: Wrap docs at 79 chars.

Author: Fixed. I had to keep 
```
    class SomeForm(forms.Form):
        country = forms.ChoiceField(choices=[c.name for c in Country.objects.all()])
```
over 79 chars though, because blacken-docs was yelling at me ☹️

  :setting:`INSTALLED_APPS` to contain
  ``'django.contrib.admin.apps.SimpleAdminConfig'`` instead of
  ``'django.contrib.admin'``.

> * ``RuntimeWarning: Accessing the database during app initialization is
>   discouraged.`` This happens if one of your apps attempts database access in
>   either the :meth:`AppConfig.ready` method or during import time. This is bad
>   because it will happen during the startup of every management command.

Reviewer 2: Rewording + adding more reasons:

```suggestion
* ``RuntimeWarning: Accessing the database during app initialization is
  discouraged.`` This warning is triggered for database queries that run before
  apps are ready, such as during module imports. Such premature database
  queries are discouraged because they will slow down your project startup,
  potentially cache stale data, and can even fail if migrations are pending.
```

I even helped a client last week who had an import-time database query that started failing after they added a migration. All their processes failed, including `./manage.py migrate`, so they couldn't actually apply the migration without a workaround!

Author: I kept the reference to `AppConfig.ready()` and the part that the queries run at the startup of every management command if that's okay.

Reviewer 2: Y“During the startup of every management command“ is true, but slightly less than complete, because processes that load the project for you, like gunicorn and pytest, aren’t management commands but are still affected. This is why I suggested the generic “your project startup“.

I’m happy to leave as “every management command“, but perhaps there’s other wording like “every process“ or “every command“? I'm not sure what's clearest.

Author: Since the warning block below `AppConfig.ready()` also talks about management commands, I would also leave it as "management commands" here. Unless you think we should rewrite the warning block as well.

https://github.com/django/django/blob/f06eafa94744229b1e44c2029aa3a1f4cc826adf/docs/ref/applications.txt#L301-L312

Reviewer 2: Yeah okay, let's stick with it.

>   For example, a common mistake may be to populate the choices of a form field
>   like this::

Reviewer 2: Let's not hedge, I think we can assert it's fairly common:

```suggestion
  For example, a common mistake is making a database query to populate form
  field choices::
```

>     class SomeForm(forms.Form):

Reviewer 2: Let's use a more realistic name.
```suggestion
    class LocationForm(forms.Form):
```

        country = forms.ChoiceField(choices=[c.name for c in Country.objects.all()])

>   In the example above, the ``Country.objects.all()`` query would be executed
>   as soon as the module containing this form would be imported. To silence the
>   warning, use a :class:`~django.forms.ModelChoiceField` instead::

Reviewer 2: "silence the warning" typically implies ignoring it - instead, we're avoiding/fixing it. Attempted wording improvement:

```suggestion
  In the example above, the query from ``Country.objects.all()`` is executed
  when the module imports, because the ``QuerySet`` is iterated over. To avoid
  the warning, the form could use a :class:`~django.forms.ModelChoiceField`
  instead::
```

    class SomeForm(forms.Form):
        country = forms.ModelChoiceField(queryset=Country.objects.all())

>   To make it easier to find the code that triggered this warning, tell Python
>   to :ref:`treat warnings as errors <python:warning-filter>` using, for
>   example, ``python -Werror manage.py shell`` which would give you a complete
>   stack trace.

Reviewer 2: Rewording to group the "as errors" and "stack trace" concepts before giving an example.

```suggestion
  To make it easier to find the code that triggered this warning, you can make
  Python :ref:`treat warnings as errors <python:warning-filter>` to reveal the
  stack trace, for example with ``python -Werror manage.py shell``.
```

  :setting:`INSTALLED_APPS` to contain
  ``'django.contrib.admin.apps.SimpleAdminConfig'`` instead of
  ``'django.contrib.admin'``.

* ``RuntimeWarning: Accessing the database during app initialization is

>   discouraged.`` This happens if one of your apps accesses the databases in the

Reviewer 4: ```suggestion
  discouraged.`` This happens if one of your apps attempts database access in the
```
I was going to suggest removing `s` from `databases` but `attempts datbase access` can be used for 1 or many dbs.

Reviewer 3: I'd prefer David's version with `database access` instead of `SQL queries` as it's closer to the raised exception and more generic.

>   :meth:`~AppConfig.ready` method or directly at import time. This is bad

Reviewer 4: ```suggestion
  :meth:`~AppConfig.ready` method or during import time. This is bad
```

  because it will happen during the startup of every management command. For
  example, even though the test database configuration is separate from the
  production settings, ``manage.py test`` would still execute some queries

>   against your **production** database!

Reviewer 4: Not a useful/common example? 🤔

Instead of focusing on one speculative example, consider describing the problem generally. For eg:

```
Database access during app initalisation or import time is undesirable for a number of reasons and is never required in Django projects.

For example a common mistake may be to populate the choices on a form field …
```

Author: Fair enough. I just copied from the warning note in the `AppConfig.ready()` section 😅
